### PR TITLE
Added macos support

### PR DESCRIPTION
**Pull request by levenon**
Added macos_framework and macos_static_framework to support macos.

### CODE REVIEW

**Comment by brentleyjones** [comment]
Can you add some tests for these new rules, to verify they work, and that we don't regress them?

**Comment by luispadron** [comment]
@levenon Is this something you're working on updating or would you need support with? We'd love these supported in rules_apple for some macOS work we have planned.

**Comment by levenon** [comment]
> 



> Can you add some tests for these new rules, to verify they work, and that we don't regress them?

ok! i will

**Comment by levenon** [comment]
> @levenon Is this something you're working on updating or would you need support with? We'd love these supported in rules_apple for some macOS work we have planned.

We need it for cross-platform purposes

**Comment by luispadron** [comment]
I think this is ready for review as well, thanks for adding tests @levenon!

File: apple/internal/rule_support.bzl [hunk 1]
```diff
@@ -553,6 +553,7 @@ _RULE_TYPE_DESCRIPTORS = {
                 # XPC Application binaries live in
                 # Application.app/Contents/XCPServices/XPCService.xpc/Contents/MacOS/XPCService
                 # Frameworks are packaged in Application.app/Contents/Frameworks
+                "@executable_path/../Frameworks",
```
Comment (brentleyjones) [inline apple/internal/rule_support.bzl:556]: Is this correct?

Comment (levenon) [reply apple/internal/rule_support.bzl:556]: yes,it's correct, i updated annotations

File: test/starlark_tests/resources/assets_macos.xcassets/star.imageset/star_2x.png
[the remark this answers is not in the document]
Comment (levenon) [reply test/starlark_tests/resources/assets_macos.xcassets/star.imageset/star_2x.png:1]: [apple_resource_group](https://github.com/bazelbuild/rules_apple/blob/440015d4a1c9d7ac9a73b006e59362eebc02356f/doc/rules-resources.md#apple_resource_group) has already supported it. 
Is there anything else we need to do？

[the remark this answers is not in the document]
Comment (brentleyjones) [reply test/starlark_tests/resources/assets_macos.xcassets/star.imageset/star_2x.png:1]: What I mean is, is there a way to test what ever you are trying to test without checking in more images into the repo?

[the remark this answers is not in the document]
Comment (levenon) [reply test/starlark_tests/resources/assets_macos.xcassets/star.imageset/star_2x.png:1]: yes, i'v merged these images in a single imageset

**Review by brentleyjones — APPROVED** [review]

File: apple/internal/macos_rules.bzl [hunk 2]
```diff
@@ -83,30 +83,59 @@ load(
...
-    embedded_targets = ctx.attr.extensions + ctx.attr.xpc_services
-
     actions = ctx.actions
     apple_mac_toolchain_info = ctx.attr._mac_toolchain[AppleMacToolsToolchainInfo]
     apple_xplat_toolchain_info = ctx.attr._xplat_toolchain[AppleXPlatToolsToolchainInfo]
     bundle_id = ctx.attr.bundle_id
     bundle_name, bundle_extension = bundling_support.bundle_full_name_from_rule_ctx(ctx)
     executable_name = bundling_support.executable_name(ctx)
-    bundle_verification_targets = [struct(target = ext) for ext in embedded_targets]
+    bundle_verification_targets = [struct(target = ext) for ext in ctx.attr.extensions]
```
Comment (thii) [inline apple/internal/macos_rules.bzl:131]: The original `bundle_verification_targets` also contain `ctx.attr.xpc_services`. Why was that removed?

File: apple/internal/macos_rules.bzl [hunk 3]
```diff
@@ -1827,3 +1864,675 @@ macos_dylib = rule_factory.create_apple_binary_rule(
...
+        DefaultInfo(files = processor_result.output_files),
+        MacosStaticFrameworkBundleInfo(),
+        OutputGroupInfo(**processor_result.output_groups),
+    ] + processor_result.providers
+
+macos_framework = rule_factory.create_apple_bundling_rule(
+    implementation = _macos_framework_impl,
+    platform_type = "macos",
+    product_type = apple_product_type.framework,
+    doc = """Builds and bundles an macOS Dynamic Framework.
```
Comment (thii) [inline apple/internal/macos_rules.bzl:2484]:
```suggestion
    doc = """Builds and bundles a macOS Dynamic Framework.
```

File: apple/internal/macos_rules.bzl [hunk 4]
```diff
@@ -1827,3 +1864,675 @@ macos_dylib = rule_factory.create_apple_binary_rule(
...
+
+To use this framework for your app and extensions, list it in the `frameworks` attributes
+of those `macos_application` and/or `macos_extension` rules.""",
+)
+
+macos_dynamic_framework = rule_factory.create_apple_bundling_rule(
+    implementation = _macos_dynamic_framework_impl,
+    platform_type = "macos",
+    product_type = apple_product_type.framework,
+    doc = "Builds and bundles an macOS dynamic framework that is consumable by Xcode.",
```
Comment (thii) [inline apple/internal/macos_rules.bzl:2494]:
```suggestion
    doc = "Builds and bundles a macOS dynamic framework that is consumable by Xcode.",
```

File: apple/internal/macos_rules.bzl [hunk 5]
```diff
@@ -1827,3 +1864,675 @@ macos_dylib = rule_factory.create_apple_binary_rule(
...
+)
+
+macos_static_framework = rule_factory.create_apple_bundling_rule(
+    implementation = _macos_static_framework_impl,
+    platform_type = "macos",
+    product_type = apple_product_type.static_framework,
+    doc = """Builds and bundles an macOS static framework for third-party distribution.
```
Comment (thii) [inline apple/internal/macos_rules.bzl:2501]:
```suggestion
    doc = """Builds and bundles a macOS static framework for third-party distribution.
```

File: apple/internal/rule_factory.bzl [hunk 6]
```diff
@@ -832,7 +833,65 @@ set, then the default extension is determined by the application's product_type.
...
+A list of library targets on which this framework depends in order to compile, but the transitive
+closure of which will not be linked into the framework's binary.
+""",
+            ),
+            "exclude_resources": attr.bool(
+                default = False,
+                doc = """
+Indicates whether resources should be excluded from the bundle. This can be used to avoid
+unnecessarily bundling resources if the static framework is being distributed in a different
+fashion, such as a Cocoapod.
```
Comment (thii) [inline apple/internal/rule_factory.bzl:891]:
```suggestion
fashion, such as CocoaPods.
```

File: apple/internal/macos_rules.bzl [hunk 7]
```diff
@@ -216,7 +246,7 @@ def _macos_application_impl(ctx):
             platform_prerequisites = platform_prerequisites,
             provisioning_profile = provisioning_profile,
             rule_descriptor = rule_descriptor,
-            targets = ctx.attr.deps + embedded_targets,
+            targets = ctx.attr.deps + ctx.attr.extensions + ctx.attr.frameworks,
```
Comment (thii) [inline apple/internal/macos_rules.bzl:249]: I think this should be kept the same? Otherwise `ctx.attr.xpc_services` would be left out as well.

File: apple/internal/macos_rules.bzl [hunk 8]
```diff
@@ -251,7 +282,7 @@ def _macos_application_impl(ctx):
         partials.apple_symbols_file_partial(
             actions = actions,
             binary_artifact = binary_artifact,
-            dependency_targets = embedded_targets,
+            dependency_targets = embedded_targets + ctx.attr.deps,
```
Comment (thii) [inline apple/internal/macos_rules.bzl:285]: `embedded_targets` already include `ctx.attr.deps`.
